Allow return to be omitted from single expression functions.

nate-chandler opened the pull request:
Enables return to be omitted from function bodies which consist of only a single expression.

nate-chandler committed:
WIP: Allow return to be omitted from single expression functions.

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler committed:
Used a constraint system, not typeCheckExpression.

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler committed:
Just revert return insertion if inappropriate.

Because the BodyStmts are revered to their original contents, they are
no longer implicit.

nate-chandler commented:
@swift-ci please test source compatibility

jrose-apple commented on lib/AST/ASTWalker.cpp:
```diff
@@ -330,6 +330,15 @@ class Traversal : public ASTVisitor<Traversal, Expr*, Stmt*,
+    //         AFD->setSingleExpressionBody(body);
+    //       else
+    //         return true;
+    //     }
```
Maybe continue processing the body the old way, but check if it goes from single-expression to non-single-expression? I think PlaygroundTransform might do that (because it will insert logging).

nate-chandler replied:
Is this the kind of thing that you had in mind?

```
    // Check whether a function which was single-expression is no longer because
    // of code being inserted.
    if (AFD->hasSingleExpressionBody()) {
      bool stillHasSingleExpressionBody = true;
      if (auto *BS = AFD->getBody()) {
        switch (BS->getNumElements()) {
          case 0: {
            stillHasSingleExpressionBody = false;
            break;
          }
          case 1: {
            auto element = BS->getElement(0);
            if (auto stmt = element.dyn_cast<Stmt *>()) {
              if (!isa<ReturnStmt>(stmt) && !isa<FailStmt>(stmt)) {
                stillHasSingleExpressionBody = false;
                break;
              }
            } else if (element.is<Decl *>()) {
              stillHasSingleExpressionBody = false;
            }
            break;
          }
          case 2: {
            if (!BS->getElement(0).is<Expr *>()) {
              stillHasSingleExpressionBody = false;
              break;
            }
            if (auto *stmt = BS->getElement(1).dyn_cast<Stmt *>()) {
              if (!isa<ReturnStmt>(stmt)) {
                stillHasSingleExpressionBody = false;
              }
              break;
            } else {
              stillHasSingleExpressionBody = false;
              break;
            }
          }
        }
      }
      if (!stillHasSingleExpressionBody) {
        AFD->setHasSingleExpressionBody(false);
      }
    }
```

jrose-apple replied:
:-( Yeah, this is pretty nasty. I don't know what to do here. 

(I do suggest writing PlaygroundTransform tests and seeing how they handle all this.)

nate-chandler replied:
Okay, several tests added here: 
https://github.com/apple/swift/pull/23251/commits/a7daf25f74f42b49414e459bb5b45dc8630902dd .  They look to behave correctly, so maybe we don't need to do anything here after all?

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -445,6 +445,26 @@ case DeclKind::ID: return cast<ID##Decl>(this)->getLoc();
+  assert(braceStmt != nullptr && "No body currently available.");
+  auto body = getBody()->getElement(0);
+  if (body.is<Stmt *>())
+    return cast<ReturnStmt>(body.get<Stmt *>())->getResult();
```
Nitpick (here and elsewhere): We'd usually write this as

```
if (auto *stmt = body.dyn_cast<Stmt *>())
  return cast<ReturnStmt>(stmt)->getResult();
```

to avoid the `is`/`get` repetition.

jrose-apple replied:
Can we get into the state where it's an expression, though? It seems like the way you have your change set up, the parser will always produce a ReturnStmt.

nate-chandler replied:
Fixed up all the casting to use `dyn_cast` instead of the repetitive use of `is` and `get` in 
https://github.com/apple/swift/pull/23251/commits/a310279dabda30f1de50b2a6bb2c11c0ef20868f .

nate-chandler replied:
I think we can get into that state after the second modification of the body happens, because the return that got added gets removed again, right?

jrose-apple replied:
At that point it's not a single-expression closure, right?

nate-chandler replied:
I'm not sure.  Looking at the way that this is done for closures (`ExprWalker::walkToExprPre`, `ExprRewriter::coerceClosureExprFromNever` and `ExprRewriter::coerceClosureExprToVoid`), it looks like the closures remain single expression even though their bodies are rewritten not to have the return (in the case of `coerceClosureExprFromNever`) or two have the original expression followed by a `return ()` (in the case of `coerceClosureExprToVoid`).

jrose-apple replied:
Ah, I see. The flag is more about how it was written in source, then. Okay, that makes sense.

jrose-apple commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -5621,8 +5621,56 @@ void Parser::parseAbstractFunctionBody(AbstractFunctionDecl *AFD) {
+      auto Element = BS->getElement(0);
+      // FIXME: FIXME_NOW: DETERMINE: Do we need to do this?  It's being done
+      //        for closures (Parser::parseExprClosure) but it may not be 
+      //        relevant here.
```
I think it's not currently relevant because we don't infer return types, but it would still be nice to be uniform.

nate-chandler replied:
Okay, I uncommented and cleaned this up in 
https://github.com/apple/swift/pull/23251/commits/a310279dabda30f1de50b2a6bb2c11c0ef20868f .

jrose-apple commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -5621,8 +5621,56 @@ void Parser::parseAbstractFunctionBody(AbstractFunctionDecl *AFD) {
+        } else if (auto F = dyn_cast_or_null<ConstructorDecl>(AFD)) {
+          if (F->getFailability() != OTK_None && isa<NilLiteralExpr>(E)) {
+            // If it's a nil literal, just insert return.  This is the only 
+            // legal thing to return.
```
\*sigh\* I guess this makes sense.

```swift
init?() { nil }
```

nate-chandler replied:
:-D What about `init?() { try nil }`

jrose-apple commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -5621,8 +5621,56 @@ void Parser::parseAbstractFunctionBody(AbstractFunctionDecl *AFD) {
+      //           } else 
+      if (Element.is<Expr *>()) {
+        auto E = Element.get<Expr *>();
+        if (auto F = dyn_cast_or_null<FuncDecl>(AFD)) {
```
Nitpick: The `or_null` variant is for when the value might be null, not what you want as output. `dyn_cast` is fine here.

nate-chandler replied:
Ah ha!  These usages of `dyn_cast_or_null` are gone in 
https://github.com/apple/swift/pull/23251/commits/a310279dabda30f1de50b2a6bb2c11c0ef20868f .

jrose-apple commented on test/Parse/omit_return.swift:
```diff
@@ -0,0 +1,1408 @@
+
+var fvg_missing: String {
+    get {
+    }
```
Shouldn't this produce *some* diagnostic? Or does that not happen until SIL diagnostics, because it's dataflow-dependent?

nate-chandler replied:
Yup!  Compiling the code with the debug build of these changes produces the error you're expecting:

```
error: missing return in a function expected to return 'String'
    }
    ^

```

jrose-apple commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -1673,6 +1674,52 @@ bool TypeChecker::typeCheckFunctionBodyUntil(FuncDecl *FD,
+  //        live?
+  if (FD->hasSingleExpressionBody()) {
+    auto fnTypeLoc = FD->getBodyResultTypeLoc();
+    auto fnType = fnTypeLoc.getType();
```
Nitpick: `fnType` is a confusing name for the result type.

nate-chandler replied:
Fixed in 
https://github.com/apple/swift/commit/a310279dabda30f1de50b2a6bb2c11c0ef20868f where that variable is eliminated and the other, `fnTypeLoc` is renamed to `resultTypeLoc`.

jrose-apple commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -1673,6 +1674,52 @@ bool TypeChecker::typeCheckFunctionBodyUntil(FuncDecl *FD,
+    auto fnTypeLoc = FD->getBodyResultTypeLoc();
+    auto fnType = fnTypeLoc.getType();
+    auto returnStmt = cast<ReturnStmt>(BS->getElement(0).get<Stmt *>());
+    auto E = returnStmt->getResult();
```
Nitpick: shouldn't this use the accessor for `hasSingleExpressionBody`?

nate-chandler replied:
Yup!  Fixed in 
https://github.com/apple/swift/pull/23251/commits/a310279dabda30f1de50b2a6bb2c11c0ef20868f .

jrose-apple commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -1673,6 +1674,52 @@ bool TypeChecker::typeCheckFunctionBodyUntil(FuncDecl *FD,
+    //             FD->setSingleExpressionBody(E);
+    //           }
+
+    if (fnTypeLoc.isNull() || fnType->isVoid()) {
```
You've already asked for `getType()`, so you know the TypeLoc isn't null.

nate-chandler replied:
Fixed so that `getType()` isn't called before checking that `fnTypeLoc.isNull()` is false.

jrose-apple commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -1673,6 +1674,52 @@ bool TypeChecker::typeCheckFunctionBodyUntil(FuncDecl *FD,
+                             /*implicit=*/false);
+    } else {
+      // FIXME: FIXME_NOW: DETERMINE: Is this the appropriate mechanism to use
+      //        to divine the type of E?
```
Not quite, because it might be context-sensitive:

```swift
var foo: String { return .init() }
var bar: Int8 { return 0 }
```

Besides that it seems mostly reasonable, but I'll have to double-check, or get someone more familiar with the type checker to answer.

nate-chandler replied:
It looks like these two cases work.  You inspired me to add ( 
https://github.com/apple/swift/pull/23251/commits/a5f94069a63f7a144de20e39113a3ec721bea52d ) a lot more tests, though, with functions that implicitly return a number of additional classes of expr.  There are a few implicit exprs for which the implicit return triggers assertion failures: `CoerceExpr`, `ConditionalCheckedCastExpr`, `ForcedCheckedCastExpr`, `IsExpr`, and `ConditionalExpr`: `Assertion failed: (!ifExpr->isFolded() && "already folded if expr in sequence?!")`  If I understand the assertion failure and the effect of `preCheckExpression` correctly, the problem is that `preCheckExpr` folds the sequence and then calling `typeCheckBody` at the end tries to fold it again.  But it looks like `preCheckExpr` has to be called before solving the constraint system.

nate-chandler replied:
Caching the result (a bit for whether it's been preChecked and a pointer to the replacement expr if it has) of `preCheckExpr` on the Expr being prechecked avoids that assert.  Does that sound like a reasonable solution, @jrose-apple?

```
/// Pre-check the expression, validating any types that occur in the
/// expression and folding sequence expressions.
bool TypeChecker::preCheckExpression(Expr *&expr, DeclContext *dc) {
  if (expr->getIsPreChecked()) {
    auto result = expr->getPreCheckResult();
    expr = result;
    return result == nullptr;
  }
  PreCheckExpression preCheck(*this, dc, expr);
  // Perform the pre-check.
  if (auto result = expr->walk(preCheck)) {
    expr->setPreChecked(result);
    expr = result;
    return false;
  }
  expr->setPreChecked(nullptr);
  return true;
}
```

jrose-apple replied:
:-( I don't like that we don't know if it's going to happen or not. Can we go the other way instead, where we just say single-expression functions never go down the normal path, and just call typeCheckStatement explicitly if they end up making a new statement? That's closer to what we're doing for closures.

nate-chandler replied:
I was overlooking the fact that `preCheckExpression` was handing back the modified (potentially folded) expression.  Writing the updated expression back into the single expression body field of the function resolves the refolding assertion.  This feels a lot better to me than what I was proposing above.  

The type-checking does still occur twice, but we can use the expression's type to determine whether or not to remove the return, just like the closure case is doing.

```
// CSApply.cpp:7522
if (cs.getType(body)->isUninhabited()) {
  closure = Rewriter.coerceClosureExprFromNever(closure);
}
```

```
// CSApply.cpp:5864
auto returnStmt = cast<ReturnStmt>(member.get<Stmt *>());
auto singleExpr = returnStmt->getResult();

cs.setExprTypes(singleExpr);
tc.checkIgnoredExpr(singleExpr);

SmallVector<ASTNode, 1> elements;
elements.push_back(singleExpr);

auto braceStmt = BraceStmt::create(tc.Context,
                                   closureExpr->getStartLoc(),
                                   elements,
                                   closureExpr->getEndLoc(),
                                   /*implicit*/true);
```

What do you think about this approach?

jrose-apple replied:
> The type-checking does still occur twice, but we can use the expression's type to determine whether or not to remove the return, just like the closure case is doing.

I don't understand this bit. In the closure case, we have an extra conversion between the type of the closure literal and the expected type, and that's where the coercion happens. We don't have that here, where you're just type-checking the plain expression (per the other comment with a preferred overload).

nate-chandler replied:
I was trying to highlight the similarity between this approach and what's being done for closures, but I recognize the difference you describe.  I'll see about changing the implementation to align more closely with the approach taken for closures.

jrose-apple commented:
Forgot the summary at the top of "This looks pretty good; I'm glad it doesn't seem to take too many changes; good job listing out test cases."

Some other things to check:
- code completion (as you mentioned)
- what these look like when using SwiftSyntax

nate-chandler committed:
Just reuse old body when reverting return insertion.

nate-chandler committed:
Addressed issues from PR review.

- preferred dyn_cast to is and get
- restored insertion of () on bare return statements
- used dyn_cast not dyn_cast_or_null when the argument is non-null
- used getSingleExpressionBody accessor during second AST modification
- eliminated erroneous access through null reference

nate-chandler committed:
Tested implicit return of most exprs from free functions.

jrose-apple commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -1673,6 +1674,42 @@ bool TypeChecker::typeCheckFunctionBodyUntil(FuncDecl *FD,
+      SmallVector<constraints::Solution, 4> viable;
+      if(!cs.solve(E, /*convertType*/Type(), nullptr, viable,
+                   FreeTypeVariableBinding::Disallow)) {
+        auto &solution = viable[0];
```
This still weirds me out. How do we know there's a solution?

nate-chandler replied:
If I understand correctly, solve will return true if we don't get any solutions:

```
/// \returns true is an error occurred, false is system is consistent
/// and solutions were found.
```

jrose-apple replied:
\>< That's what I get for not reading the docs. I think there's still a *small* hole here where we have overloads that may or may not be the same as the result type, which means multiple solutions. In that case the contextual type will make a difference.

nate-chandler replied:
Is this the kind of scenario you were thinking of, where overloads with different return types would mean multiple solutions?

```
enum MyOwnNever {}
func die() -> MyOwnNever { print("MyOwnNever"); fatalError() }
func die() -> Never { print("Never"); fatalError() }
func die() -> Int { 6 }

func roll() -> Int { die() }
func roll2() -> Never { die() }
func roll3() -> MyOwnNever { die() }
```

It turns out that it works (and that `viable.size()` is always 1).  Could it be that the providing `FD` as the `DeclContext` to the `ConstraintSystem` is providing the context necessary to disambiguate?

jrose-apple replied:
> Is this the kind of scenario you were thinking of, where overloads with different return types would mean multiple solutions?

Not quite. There's a scenario where one choice *is* better (because of other overload resolution rules), but it's not the "right" one.

```
func evil() -> Never { print("Never"); fatalError() }
func evil(extraArgument: Int = 0) -> Int { return extraArgument }

func roll() -> Int { evil() }
```

However…I'm realizing that interpreting this as `return evil()` would be a source-breaking change, which means we have to make a choice as to whether that's allowed. So one way to go here is a funny extra rule in single-expression functions that's different from single-expression closures (possibly conditional on language mode): "unlike with an explicit `return`, overload resolution does not take the calling function's return type into account; it's only implied if the types match up".

(I have to eat crow now because in a previous discussion with other Apple folks I said "there exists an implementation that's not source-breaking at all", and there does, but it's unfortunate that that implementation isn't going to follow the same rules as we get for closures.)

> Could it be that the providing `FD` as the `DeclContext` to the `ConstraintSystem` is providing the context necessary to disambiguate?

I don't think so. I guess I'll have to bother a type-checker person, or one of us has to track down why it's working.

nate-chandler replied:
Does the source compatibility breakage only happen if the preferred overload returns `Never`?  If so, could the difference between implicit returns for functions and closures be specific to overload resolution when one of the candidates returns `Never`?

It would be nice to keep support for

```
func zero() -> String { "" }
func zero(plus other: Int = 0) -> Int { other }
func identity() -> Int { zero() }
```

where none of the candidates return Never since selecting the one preferred by overload resolution rules wouldn't have been legal code without implicit return.

That would make the rule even more complicated, but maybe it would be worth it?  Overloading where one of the candidates returns `Never` might be less common (than overloading where none of the candidates returns `Never`) enough that it would be worth the rule for that case being different.

jrose-apple replied:
`Never` or something `Never`-ish, yeah. But that seems harder to implement if you go the direction I suggest where we manage to only type-check the expression one way.

nate-chandler replied:
If we stick with the two-pass approach (with writing back the prechecked expression back into the single expression body of the function), it seems like we get the best of both worlds.  

I looked at a few permutations of your example:

```
func evil_1(value: Int = 0) -> Int { return 0 }
func evil_1() -> Never { fatalError() }

func pass_1_noreturn() -> Int { evil_1() }
func pass_1_return() -> Int { return evil_1() }

func evil_2() -> Int { return 0 }
func evil_2(value: Int = 0) -> Never { fatalError() }

func pass_2_noreturn() -> Int { evil_2() }
func pass_2_return() -> Int { return evil_2() }


print(pass_1_return())
print(pass_2_return())
print(pass_2_noreturn())
print(pass_1_noreturn())
```

Without implicit return:
pass_1_noreturn: traps
pass_1_return: returns 0
pass_2_noreturn: fails to compile (`error: missing return in a function expected to return 'Int'`)
pass_2_return: returns 0

With implicit return:
pass_1_noreturn: traps
pass_1_return: returns 0
pass_2_noreturn: returns 0
pass_2_return: returns 0

If I'm looking at this right, this seems to indicate that we get source compatibility without having to special case the scenario where the overload set includes a function which returns Never.  Is that right?

jrose-apple replied:
That seems correct. I just don't like the two passes. Is it possible to say "if the expression type is uninhabited, write it back *and then skip checking the body?*" That way we're never type-checking the output of a type-check.

…And even with that, for performance reasons I don't like type-checking the same expression twice when there's not an error. Can we go the other way instead, and say "try to type-check this as if we're converting it to the return type, and if *that* fails try to type-check it as producing an uninhabited type, and if *that* fails print the diagnostics from the first system"?

nate-chandler replied:
That makes a lot of sense, I'll try that out.

nate-chandler replied:
Okay, as of 
https://github.com/apple/swift/pull/23251/commits/a1023f2aa475d4ad8d606ae3d7dca7d0d8dd8d86 , it solves the constraint system which the expression to be returned.  If that succeeds, the body is updated and the function returns.  Failing to solve that system, it tries to solve a second system where the expression is not returned.  If that succeeds, the expression is uninhabited and the expression is unequal to the function's return type, it deletes the return, updates the body, and returns.  If that fails, the diagnostics from attempting to salvage the first system are shown.  How does that compare with what you had in mind?

jrose-apple replied:
Oof. That's exactly what I asked for and then I thought it through again and doesn't it break source compatibility in your `pass_1_noreturn` case above? That really is a case where there are multiple solutions and we need to pick the unusual one in order to not break source compatibility.

nate-chandler replied:
Yeah, it does.  

We could just reorder the two cases, checking first whether we have an uninhabited solution to the system without a specified conversion type and then checking whether we have any solution to the system with a specified conversion type.  That unbreaks the `pass_1_noreturn` case.

It does, unfortunately, put us back into have two passes most of the time.  We can at least avoid type checking twice in the scenario where the expression is uninhabited, which gives the solution you mentioned above

> Is it possible to say "if the expression type is uninhabited, write it back and then skip checking the body?" That way we're never type-checking the output of a type-check.

How does that sound?

jrose-apple replied:
:-( I don't think we can take a feature that requires type-checking something twice. That's still potentially a major slowdown if it's a really complex expression that *might* produce Never (but probably won't).

jrose-apple replied:
I'll ask Doug and/or Pavel what they think tomorrow.

nate-chandler replied:
That makes sense.  I'm going to look into the possibility of special casing overload resolution for single-expression functions.

jrose-apple replied:
The specific special case I would think to try is "allow coercing T to the contextual type if T is uninhabited".

nate-chandler replied:
A couple more data points:

"Allowing coercion from T to the contextual type if T is uninhabited" 

```
// ExprRewriter::coerceToType
  case TypeKind::Enum: {
    auto fromEnum = cast<EnumType>(desugaredFromType);
    if (fromEnum->isUninhabited()) {
      return cs.cacheType(new (tc.Context) UninhabitedUpcastExpr(expr, toType));
    }
    break;
```

requires that `solve` not be provided the function's result type as its `convertType`.  

```
// TypeChecker::typeCheckFunctionBodyUntil
      constraints::ConstraintSystem cs(*this, FD,
                                       constraints::ConstraintSystemFlags::SuppressDiagnostics);
      cs.setContextualType(E, TypeLoc::withoutLoc(resultType), CTP_ReturnStmt);

      SmallVector<constraints::Solution, 4> viable;
      if(!cs.solve(E, /*convertType*/Type(), /*listener=*/nullptr, viable,
                    FreeTypeVariableBinding::Disallow)) {
```

That works for those four fatalError overload cases above, but causes assertion failures when applying the solution on more basic things like

```
func wrong() -> String { 17 }
```

because the constraint system features an unhandled coercion:

```
Fixed score: 0 0 0 0 0 0 0 0 0 0 0 0
Type variables:
  $T0 as Int @ locator@0x7fb492039200 [IntegerLiteral@test2.swift:17:5]

Overload choices:

Constraint restrictions:

Disjunction choices:
Unhandled coercion
```

Taking that a step further, allowing uninhabited types to match arbitrary types

```
// ConstraintSystem::matchTypes
  if (kind >= ConstraintKind::Conversion) {
    if (type1->isUninhabited()) {
      conversionsOrFixes.push_back(ConversionRestrictionKind::UninhabitedUpcast);
    }
  }
```

(i.e. treating uninhabited types as bottom types, another language change) allows most of the new code in  `TypeChecker::typeCheckFunctionBodyUntil` to be deleted (and in particular type checking occurs once always), but continues to change the behavior of `pass_1_noreturn` if the solution is weighted:

```
ConstraintSystem::TypeMatchResult
ConstraintSystem::matchUninhabitedUpcastTypes(Type type1, Type type2,
                                              TypeMatchOptions flags,
                                              ConstraintLocatorBuilder locator) {
  if (type1->isUninhabited()) {
    increaseScore(SK_UninhabitedUpcast);
    return getTypeMatchSuccess();
  }

  assert(false);
  return getTypeMatchFailure(locator);
}
```

Dropping the weighting makes `pass_1_noreturn` behave as before but unacceptably changes the behavior of the solution of

```
func pass_1_return() -> Int { return evil_1() }
``` 

to resolve `evil_1` to the `Never` overload.

I'm looking at the feasibility of passing the fact that we're type-checking a single-expression function down into the constraint system so that we can not weight whenever return is omitted.

nate-chandler replied:
@jrose-apple  Okay, I think I was able to resolve this: https://github.com/apple/swift/pull/23251/commits/a42c2c1a702432aa4c0add85b884b294c772ebc0 .  The change allows implicit conversion from an uninhabited type, but only from the type of a single expression of a function to the function's result type.  The conversion unlocks eliminating the less pleasant parts of the `TypeChecker::typeCheckFunctionBodyUntil`, leaving just the check for `Void` and the removal of the return in that case.  Just as importantly, it means that type-checking single expression functions goes through the same path as all other functions, giving us back all the diagnostics that get inserted along the way.

benlangmuir commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -5621,8 +5621,47 @@ void Parser::parseAbstractFunctionBody(AbstractFunctionDecl *AFD) {
+    // FIXME: FIXME_NOW: DETERMINE: Do we need to handle the code completion 
+    //        case here?  It is being handled for closure 
+    //        (Parser::parseExprClosure).
+    if (BS->getNumElements() == 1) {
```
This code should do the same thing as the closure case: if `Body.hasCodeCompletion()`, it should *not* set the single-expression body bit.  You can copy the comment explaining why here as well.

FYI I'm working on a patch to take return type into account during code-completion of single-expression closures; I can do the same for functions for this PR.

nate-chandler replied:
Cool!  Okay, the check and comment are there now: 
https://github.com/apple/swift/pull/23251/commits/b5705af35ff6563103d4fcd57183a47d96ddc157 .

nate-chandler committed:
Don't insert return during code completion.

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler committed:
Placed prechecked expressions back in body.

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler committed:
Only remove return for a unique viable solution.

nate-chandler committed:
Expected errors around implicitly returning #line.

nate-chandler committed:
Avoided double typechecking when possible.

Optimized for the common case where a single expression in a function
body is an implicit return and attempt to solve the constraint system
featuring that assumption.  If that fails, attempt to solve a second
constraint system where no conversion type is provided.  If that
succeeds and the expression's type is uninhabited, then remove the
return and proceed.  If that fails, show the diagnostics from the
failure to solve the first constraint system.

nate-chandler committed:
Added comments explaining approach.

nate-chandler committed:
Visited single expression of AbstractFunctionDecl.

Taught getSingleExpressionBody to return null and
setSingleExpressionBody to require null in the case that there is a
single stmt of type FailStmt which replaces return nil during sema for
single expression failable initializers (i.e. init?() { nil }).

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler committed:
Corrected autoclosure handling.

Set autoclosure discriminators and follow visit single expression by way
of visiting brace stmt.

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler committed:
Added diagnostic for dangling expression at end.

If the final expression in a function or closure which is missing a
return has the appropriate type, rather than producing the usual
diagnostic about a missing return, produce a diagnostic with a fixit to
insert a return before thatn final expression.

h/t Nate Cook

harlanhaskins commented on test/FixCode/fixits-omit-return.swift:
```diff
@@ -0,0 +1,14 @@
+
+func ff_fixit_addreturn() -> String {
+    print("entering ff_fixit_addreturn()")
+    "foo"
```
I think this might be better served as an `expected-error` with a FixIt range specified.

```swift
// expected-error {{missing return in a function expected to return String; did you mean to return the last expression?}} {{4-4=return }}
```

nate-chandler replied:
Ah, that is much shorter!  Changed here: 
https://github.com/apple/swift/pull/23251/commits/e5947e47753a5a693ff596c3891912dde83d89b7

jrose-apple commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -1770,6 +1772,7 @@ bool TypeChecker::typeCheckFunctionBodyUntil(FuncDecl *FD,
                                /*skipClosures*/false);
           if (E != FD->getSingleExpressionBody())
             FD->setSingleExpressionBody(E);
+          setAutoClosureDiscriminators(FD, BS);
```
Are there test cases that would fail without this?

nate-chandler replied:
Yup!  Added here: 
https://github.com/apple/swift/pull/23251/commits/7fb0a208b2d5dc6dcdade1e0ff2778502361371e

nate-chandler committed:
Added tests to cover autoclosure discriminators.

nate-chandler committed:
Used expected-error to test fixit.

nate-chandler committed:
Added PlaygroundTransform tests on implicit return.

nate-chandler commented:
@swift-ci please test

nate-chandler committed:
Don't attempt to apply solution from salvage.

Match the behavior of typeCheckExpressionImpl which simply calls solve
and applySolution but never salvage.

nate-chandler commented:
@swift-ci please test

nate-chandler committed:
Implicitly convert single exprs from uninhabited.

When type-checking a return statement's result, pass a new
ContextualTypePurpose when that return statement appears in a function
with a single expression.  When solving the corresponding constraint
system, the conversion constraint will have a new ConstraintKind.  When
matching types, check whether the constraint kind is this new kind,
meaning that the constraint is between a function's single expression
and the function's result.  If it is, allow a conversion from
an uninhabited type (the expression's type) to anything (the function's
result type) by adding an uninhabited upcast restriction to the vector
of conversions.  Finally, at coercion time, upon encountering this
restriction, call coerceUninhabited, replacing the original expression
with an UninhabitedUpcastExpr.  Finally, at SILGen time, treat this
UninhabitedUpcastExpr as a ForcedCheckedCastExpr.

Eliminates the bare ConstraintSystem usage from
typeCheckFunctionBodyUntil, ensuring that the same code path is followed
for all function bodies.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler committed:
Tested implicit non/conversion from uninhabited.

nate-chandler committed:
Don't add an implicit () after return for initializers.

nate-chandler committed:
Corrected tests by removing implicit returns.

nate-chandler committed:
Replaced superseded test.

nate-chandler committed:
Tweaked remaining failing tests.

Modified so that single-expression implicit return does not throw off
tests.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler committed:
Fixed up return omission test cases for Linux.

Added objc_interop requirement to tests which require it and moved one
test into file stating that requirement.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
Both compatibility tests have this result although their statuses are reported as failed:

```
========================================
UPasses:
  UPASS: https://bugs.swift.org/browse/SR-8234, Lark, 4.0, 7604f9, Swift Package
========================================
Action Summary:
     Passed: 261
     Failed: 0
    XFailed: 27
    UPassed: 1
      Total: 289
========================================
Repository Summary:
      Total: 79
========================================
Result: UPASS
========================================
```

nate-chandler commented:
@swift-ci please test source compatibility

harlanhaskins commented:
@nate-chandler I'm un-XFAILing Lark here: https://github.com/apple/swift-source-compat-suite/pull/330

nate-chandler commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please build toolchain

nate-chandler commented:
Toolchains are available here:

macOS: https://ci.swift.org/job/swift-PR-toolchain-osx/266/artifact/branch-master/swift-PR-23251-266-osx.tar.gz
Linux: https://ci.swift.org/job/swift-PR-toolchain-Linux/207/artifact/branch-master/swift-PR-23251-207-ubuntu16.04.tar.gz

jrose-apple commented:
I'm a little concerned by the new changes to the constraint solver, since that could change the behavior of existing closures. @DougGregor or @xedin should review those; I don't think I'm qualified.

nate-chandler commented:
> I'm a little concerned by the new changes to the constraint solver, since that could change the behavior of existing closures.

I'm hoping that the changes to the constraint solver only affect single expression functions because only in that context is the contextual type purpose `CTP_ReturnSingleExpr`, the only way to form a constraint of kind `SingleExpressionFunctionReturnConversion` is when that is the contextual type purpose, and only when that is the constraint kind does an implicit conversion from uninhabited occur.  I'm hoping that means that closures are unaffected.

jrose-apple commented:
I guess that's convincing. It's a little funny to me that we *wouldn't* use CTP_ReturnSingleExpr for closures too, or try to share more there, but we don't want to do it *now.* I'm just a little sad we had to change the constraint solver at all when your earlier implementation made it look like we wouldn't have to.

nate-chandler commented:
> I'm just a little sad we had to change the constraint solver at all when your earlier implementation made it look like we wouldn't have to.

Yeah, I know what you mean.  That earlier implementation did, though, have to type check twice in some cases, and had quite a few problems with diagnostics because it was bypassing the codepaths where they were emitted between `TypeChecker::typeCheckFunctionBodyUntil` and `ConstraintSystem::solve`.

xedin commented:
@nate-chandler Can you please explain what problem you are trying to solve with adding new  conversion and score? If I understand this correctly this might be related to expressions like these?

```swift
func foo() -> Int { fatalError() }
```

Because `fatalError()` is going to be wrapped in `ReturnStmt`?

nate-chandler commented:
@xedin Yes, that is the exact case.

xedin commented:
@nate-chandler I feel like there might be a better solution for that case, I'll pull your changes and try to come up with a suggestion for you. I think we might be possible to do in `matchTypes` similarly how we have conversions from `() -> Never` to `() -> Void` and `() -> T` for closure result in cases like `foo() -> Int { fatalError() }`

nate-chandler commented:
@xedin Okay!  Looking forward to hearing how we can make this better.

xedin committed:
[ConstraintLocator] Add a special locator element to denote contextual type of single expr function

xedin committed:
[ConstraintSystem] Use special locator for expr representing return of single expr function

xedin commented:
@nate-chandler I've opened PR for you https://github.com/nate-chandler/swift/pull/2 which uses a special locator element instead of conversion and fixes up brace statement afterwards. This would allow you to remove conversion and other stuff from constraint system.

xedin commented on test/stmt/statements.swift:
```diff
@@ -177,6 +177,7 @@ return 42 // expected-error {{return invalid outside of a func}}
 return // expected-error {{return invalid outside of a func}}
 
 func NonVoidReturn1() -> Int {
+  _ = 0
```
I noticed that a bunch of tests have `_ = 0` added which is not ideal, if a single expression is left in the body is it going to produce different diagnostics? I think one way to work around this is to check whether a single expression is a declaration and avoid wrapping it in return in that case or wrap its result.

nate-chandler replied:
Absolutely--this is not good.  

The [change](https://github.com/apple/swift/pull/23251/commits/e30511bdb8d8b4f0adc1b75e0db66992f4e25a74#diff-a5731a934e0c085bf4b84dabd27a6ffdR5719) I just pushed corrects a number of these.  Is this what you had in mind in order to "check whether a single expression is a declaration"?  

```
if (auto SE = dyn_cast<SequenceExpr>(E->getSemanticsProvidingExpr())) {
  if (SE->getNumElements() > 0 && isa<AssignExpr>(SE->getElement(1))) {
    // This is an assignment.  We don't want to implicitly return 
    // it.
    return;
  }
}
```

It does look like a few of them are appropriate:
(1) `test/Generics/requirement_inference.swift` -- we don't want to implicitly return the call to x.p1() which returns void -- dataflow analysis would detect an error (that nothing is returned) but that's not run as part of the test
(2) `test/Sema/availability_versions.swift` -- we don't want to implicitly return the call to `functionAvailableOn10_51` which would happen if the body consisted only of that
(3) `validation-test/stdlib/AssertDiagnosticsSIL.swift` -- we don't want to implicitly return the call to `assertionFailure` which would happen if the body consisted only of that
(4) `test/SILGen/super.swift` -- we don't want the body to consist of only a call to super.classMethod() since the function is expected to return Self and super.classMethod() returns Void


The change in `test/stmt/statements.swift` is questionable and corresponds to a questionable part of the patch: https://github.com/apple/swift/pull/23251/files#diff-a5731a934e0c085bf4b84dabd27a6ffdR5708 .  In the code I linked, a () result is being added to ReturnStmt.  Something like this is being done for closures, so I had it commented out to begin with but over the course of the review it was commented in.  In view of this test failure, it seems like that code should go.  Is that right?

I'm looking into `test/decl/subscript/subscripting.swift` and `test/decl/var/static_var.swift` further to see whether it makes sense to change those tests.

xedin replied:
@nate-chandler Yes, something like that. This got me thinking, maybe we should even attempt that. Because as in cases you have mentioned, I think it makes sense make it not `_ = 0` but something reasonable, under new rules implicit return is a correct behavior. All of the single statement functions would have an implicit return added to them if the result type of the function is specified. It maybe makes sense to add tailored diagnostic for that case though which says that single expression function can't implicitly return T because result type is specified as U.
And in case of assignments say that `non-void function should return a value`.

nate-chandler replied:
@xedin To make sure I understand, the idea is that:

(1) Assignments should be implicitly returned.  This
```swift
func foo() -> Int {
    a = b
}
```
should be rewritten as
```swift
func foo() -> Int {
    return a = b
}
```

(2) There should be a diagnostic for single expression return of ().  This test should pass:
```swift
func foo() -> Int {
    a = b // expected-error {{ non-void function should return a value }}
}
```

(3) There should be a special diagnostic for the return of an implicit ().  This code should remain:  
```swift
if (!returnStmt->hasResult()) {
    auto returnExpr = TupleExpr::createEmpty(Context,
                                             SourceLoc(),
                                             SourceLoc(),
                                             /*implicit*/true);
    returnStmt->setResult(returnExpr);
    AFD->setHasSingleExpressionBody();
    AFD->setSingleExpressionBody(returnExpr);
  }
}
```
This code
```swift
func foo() -> Int {
    return
}
```
should still be rewritten as
```swift
func foo() -> Int {
    return ()
}
```
And this test should pass:
```swift
func foo() -> Int {
    return // expected-error {{ non-void function should return a value }}
}
```

(4) There should be a diagnostic for incorrect implicit return.  This test should pass:
```swift
func foo() -> String {
    17 // expected-error {{cannot convert implicitly returned single expression of type 'Int' to return type 'String'}}
}
```

We should continue to insert () as the result of return in 

(5) The tests that are modified to have a second element so as not to have an implicit return inserted (as at least `test/Generics/requirement_inference.swift`, `test/Sema/availability_versions.swift`, `validation-test/stdlib/AssertDiagnosticsSIL.swift`, and `test/SILGen/super.swift` will need to be) should use an expression other that `_ = 0`?  For example, 

I'm drawing (5) from what you wrote here

> Because as in cases you have mentioned, I think it makes sense make it not _ = 0 but something reasonable, under new rules implicit return is a correct behavior. 

but I'm not sure I'm understanding correctly.

Thank you!

nate-chandler commented:
@xedin Thank you!  I merged your PR and will eliminate the conversion and other constraint system changes.

xedin commented:
@nate-chandler No problem!

nate-chandler committed:
Eliminated conversion from uninhabited.

nate-chandler committed:
Don't implicitly return assignments.

Corrected a number of diagnostic regressions.

nate-chandler commented:
@swift-ci please test

nate-chandler committed:
Reverted unneeded test change.

nate-chandler committed:
Removed old comment.

nate-chandler committed:
Corrected off-by-one error.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci Please clean smoke test OS X platform

nate-chandler commented:
@swift-ci please clean smoke test OS X platform

nate-chandler commented:
@swift-ci please clean test OS X platform

compnerd commented:
@nate-chandler this broke the Windows test suite!

plotfi commented on test/PlaygroundTransform/implicit_return_never.swift:
```diff
@@ -0,0 +1,18 @@
+// RUN: cp %s %t/main.swift
+// RUN: %target-build-swift -Xfrontend -playground -Xfrontend -disable-playground-transform -o %t/main %t/main.swift
+// RUN: %target-codesign %t/main
+// RUN: ! %target-run %t/main --crash 2>&1 | %FileCheck -check-prefix=CRASH-CHECK %s
```
Please use not.py next time.

nate-chandler replied:
Will do!

jrose-apple replied:
I think we prefer LLVM's `not --crash`.

slavapestov commented on lib/SILOptimizer/Mandatory/DataflowDiagnostics.cpp:
```diff
@@ -42,19 +42,37 @@ static void diagnoseMissingReturn(const UnreachableInst *UI,
+  if (numElements > 0) {
+    auto element = BS->getElement(numElements - 1);
+    if (auto expr = element.dyn_cast<Expr *>()) {
+      if (expr->getType()->getCanonicalType() == ResTy->getCanonicalType()) {
```
TypeBase::isEqual() is a shortcut for comparing `getCanonicalType()s`

nate-chandler replied:
PR with a fix at https://github.com/apple/swift/pull/24546 .

slavapestov commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -5655,8 +5655,60 @@ void Parser::parseAbstractFunctionBody(AbstractFunctionDecl *AFD) {
+      auto Element = BS->getElement(0);
+      if (auto *stmt = Element.dyn_cast<Stmt *>()) {
+        auto kind = AFD->getKind();
+        if (kind == DeclKind::Var || kind == DeclKind::Subscript ||
```
AFD is an AbstractFunctionDecl so it will never have a kind that is Var or Subscript. In general we prefer to use `isa<>` instead of checking the kind directly

nate-chandler replied:
PR with a fix at https://github.com/apple/swift/pull/24546 .
